### PR TITLE
Issue #3219156: Add support for Drupal 9

### PR DESCRIPTION
**Pull request by mglaman**
## Description
Adds Drupal 9 support so we can move onto Drupal 10!

## To Test
Add this PR to https://github.com/palantirnet/drupal-rector-sandbox/tree/d9

## Drupal.org issue
https://www.drupal.org/project/rector/issues/3219156

### CODE REVIEW

File: .github/workflows/functional_test__d8_rector_examples.yml [hunk 1]
```diff
@@ -1,12 +1,12 @@
 
-name: functional_test__rector_examples
+name: functional_test__d8_rector_examples
```
Comment (mglaman) [inline .github/workflows/functional_test__d8_rector_examples.yml]: Why didn't I just make another job in this workflow?

File: src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    public function refactor(Node $node): ?Node
+    {
+        assert($node instanceof Node\Expr\MethodCall);
+        if ($this->getName($node->name) === 'drupalPostForm') {
+            // @todo: handle when there are 4 or 5 arugments for opts and ID.
+            [$path, $edit, $button] = $node->args;
+            // @todo we _must_ inject drupalGet.
+            // new Node\Expr\MethodCall($node->var, new Node\Identifier('drupalGet'), [$path])
```
Comment (mglaman) [inline src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php]: Link to https://github.com/rectorphp/rector/discussions/6538 so we can track better replacements.

File: fixtures/d9/rector_examples_updated/tests/src/Functional/DrupalPostFormTest.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /**
+     * Passing the options as a direct value.
+     */
+    public function with_options_direct() {
+        $edit = [];
+        $edit['action'] = 'action_goto_action';
+        $this->drupalPostForm('admin/config/system/actions', $edit, 'Create', ['foo' => 'bar']);
+        $this->assertSession()->statusCodeEquals(200);
```
Comment (mglaman) [inline fixtures/d9/rector_examples_updated/tests/src/Functional/DrupalPostFormTest.php]: Update assertion to:

```
        // TODO: Drupal Rector Notice: Please delete the following comment after you've made any necessary changes.
        // You must call `$this->drupalGet("admin/config/system/actions", ['foo' => 'bar']);" before submitForm
```

File: fixtures/d9/rector_examples_updated/tests/src/Functional/DrupalPostFormTest.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * Passing the options as a direct value.
+     */
+    public function with_options_as_variable() {
+        $edit = [];
+        $edit['action'] = 'action_goto_action';
+        $options = ['foo' => 'bar'];
+        $this->drupalPostForm('admin/config/system/actions', $edit, 'Create', $options);
+        $this->assertSession()->statusCodeEquals(200);
```
Comment (mglaman) [inline fixtures/d9/rector_examples_updated/tests/src/Functional/DrupalPostFormTest.php]:
```
        // TODO: Drupal Rector Notice: Please delete the following comment after you've made any necessary changes.
        // You must call `$this->drupalGet("admin/config/system/actions", $options);" before submitForm
```

File: fixtures/d9/rector_examples_updated/tests/src/Functional/DrupalPostFormTest.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /**
+     * Passing the options as a direct value.
+     */
+    public function with_html_id() {
+        $edit = [];
+        $edit['action'] = 'action_goto_action';
+        $this->drupalPostForm('admin/config/system/actions', $edit, 'Create', [], 'foo_bar_baz');
+        $this->assertSession()->statusCodeEquals(200);
```
Comment (mglaman) [inline fixtures/d9/rector_examples_updated/tests/src/Functional/DrupalPostFormTest.php]:
```
        // TODO: Drupal Rector Notice: Please delete the following comment after you've made any necessary changes.
        // You must call `$this->drupalGet("admin/config/system/actions", []);" before submitForm
        $this->submitForm($edit, 'Create', 'foo_bar_baz');
```

File: tests/src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector/UiHelperTraitDrupalPostFormRectorTest.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * @covers ::refactor
+     * @dataProvider provideData()
+     */
+    public function test(SmartFileInfo $fileInfo): void
+    {
+        $this->markTestIncomplete(
+            'Fails since BrowserTestBase cannot be resolved, will work when drupal/core added as a dev dependency.'
+        );
+        $this->doTestFileInfo($fileInfo);
```
Comment (mglaman) [inline tests/src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector/UiHelperTraitDrupalPostFormRectorTest.php]: We don't need the dev dependency. We can add a stub file, like in https://github.com/rectorphp/rector-symfony/tree/main/stubs

**Comment by mglaman** [comment]
The commits got to be a mess. I can rebase and tidy up. Or we just squash and merge

File: src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                $this->addNodeBeforeNode(
+                    $this->nodeFactory->createLocalMethodCall('drupalGet', [$path]),
+                    $node
+                );
+            } else {
+                // @phpstan-ignore-next-line
+                $this->addNodeBeforeNode(
+                    $this->nodeFactory->createLocalMethodCall('drupalGet', [$path, $options]),
+                    $node
+                );
```
Comment (mglaman) [inline src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php]: I tried returning an array of nodes before and it didn't work. But here's an example where it does: https://github.com/rectorphp/rector-src/blob/c277e7d35c3dd930660307c23ec3c74fcad337ab/rules/Defluent/Rector/Return_/DefluentReturnMethodCallRector.php#L78

Then we don't need the ignore next line for PHPStan when using a deprecated method.

Source: https://github.com/rectorphp/rector/discussions/6538#discussioncomment-945698

**Comment by agentrickard** [comment]
Successfully tested on `PHP 8 / Drupal 9.2` and `PHP 7.4 / Drupal 8.9.16` by running tests against the `examples` module set.

On the D9 / PHP 8 side, I also ran `drupal9: ddev . vendor/bin/phpunit vendor/palantirnet/drupal-rector/tests/*`     with the following results:

```
PHPUnit 9.5.6 by Sebastian Bergmann and contributors.

.......                                                             7 / 7 (100%)

Time: 00:08.977, Memory: 56.00 MB

OK (7 tests, 14 assertions)
```

I was also able to run Rector on the Drupal 8 and Drupal 9 sites with:

`drupal8: ddev . vendor/bin/rector process vendor/palantirnet/drupal-rector/fixtures/d8/rector_examples --dry-run`

File: src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php (hunk 7 above)
Comment (agentrickard) [reply src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php]: Do we need to resolve this issue before committing this?

Comment (mglaman) [reply src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php]: Yeah, I should try the fix suggested. Otherwise this cloths break in 0.12 probably

Comment (agentrickard) [reply src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php]: OK, I will hold off until that is ready. The rest seems great!

Comment (mglaman) [reply src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php]: @agentrickard I implemented a fix to avoid using the deprecated functions

**Comment by mglaman** [comment]
Test failure is due to a release in rector/rector-nette: https://github.com/rectorphp/rector-nette/commit/0227bb6f43bde75467ee2c14219713784db47afa

**Comment by mglaman** [comment]
I opened this issue: https://github.com/rectorphp/rector-nette/issues/21. We should be green when that's all resolved upstream.

**Comment by mglaman** [comment]
Looks like we'll be clear once https://github.com/rectorphp/rector-src/pull/362 finishes up :)

**Pull request merged**